src/txrequest.h

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) 2020 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_TXREQUEST_H
+#define BITCOIN_TXREQUEST_H
+
+#include <primitives/transaction.h>
+#include <uint256.h>
+
+#include <chrono>
+#include <vector>
+
+#include <stdint.h>
+
+/** Data structure to keep track of, and schedule, transaction downloads from peers.
+ *
+ * === Specification ===
+ *
+ * We keep track of which peers have announced which transactions, and use that to determine which requests
+ * should go to which peer, when, and in what order.
+ *
+ * The following information is tracked per peer/tx combination ("announcement"):
+ * - Which peer announced it (through their NodeId)
+ * - The txid or wtxid of the transaction (collectively called "txhash" in what follows)
+ * - Whether it was a tx or wtx announcement (see BIP339).
+ * - What the earliest permitted time is that that transaction can be requested from that peer (called "reqtime").
+ * - Whether it's from a "preferred" peer or not (outbound and whitelisted peers are preferred).
```

Comment:
```suggestion
 * - Whether it's from a "preferred" peer or not (outbound and noban peers are preferred).
```

style-nit: Would be good to mention the exact permission flag, because legacy whitelisted is discouraged, has been deprecated, and might be removed some time in the future.

Reply:
I'd like to keep txrequest mostly about the decision logic and data structure, while leaving net_processing responsible for the actual policy choices ("what delays are given to which peers/requests", "what timeouts are used", "when exactly is a peer considered preferred", "when exactly is a peer considered overloaded", "how many announcements can be trakced per peer").

Of course, the explanation here is ideally as comprehensive as possible and not full of "See some random comment in net_processing for the details" references. Still, do you think it would be reasonable to say instead:

> Whether it's from a "preferred" peer or not (what that means is up to the caller, but this is designed to correspond mostly to outbound peers, or others that are more trusted)

?

Reply:
Done, approximately.